lib/concurrent/atomic/read_write_lock.rb

```diff
@@ -192,7 +203,8 @@ def acquire_write_lock
     #
     # @return [Boolean] true if the lock is successfully released
     def release_write_lock
-      c = @Counter.update { |c| c-RUNNING_WRITER }
+      c = @Counter.update { |count| count-RUNNING_WRITER }
+      @WriterLock.set(nil)
```

Comment:
Does this @WriterLock need to be set before the Counter is updated? From what I can tell, setting the counter represents your linearization point; meaning that once that is set another thread could conceivably acquire the lock, set the @WriterLock to Thread.current, then *after* that we set @WriterLock to be nil. This could cause a deadlock if the thread attempted to acquire the lock multiple times without releasing it.